src/Plutus/NativeForeignConvertible.purs

```diff
@@ -0,0 +1,289 @@
+module Plutus.NativeForeignConvertible
+  ( class NativeForeignConvertible
+  , toNativeType
+  , toForeignType
+  ) where
+
+import Prelude
+import Data.Array
+import Data.Maybe (Maybe(Just, Nothing))
+import Data.Tuple (Tuple(Tuple), fst)
+import Data.UInt (UInt, fromInt, toInt, (.&.), and, (.|.), zshr, shl)
+import Data.Newtype (class Newtype, wrap, unwrap)
+import Serialization.Address (Pointer, addressBytes, addressFromBytes)
+import Serialization.Address (Address) as Foreign
+import Plutus.Types.Address (Address, AddressRecord) as Plutus
+import Types.ByteArray
+  ( ByteArray
+  , byteArrayToIntArray
+  , byteArrayFromIntArray
+  , byteArrayFromIntArrayUnsafe
+  )
+import Serialization.Hash
+  ( ed25519KeyHashFromBytes
+  , scriptHashFromBytes
+  , ed25519KeyHashToBytes
+  , scriptHashToBytes
+  )
+import Plutus.Types.Credential
+  ( Credential(PubKeyCredential, ScriptCredential)
+  , StakingCredential(StakingHash, StakingPtr)
+  )
+import Plutus.Types.AddressHeaderType
+  ( AddressHeaderType
+      ( PaymentKeyHash_StakeKeyHash
+      , ScriptHash_StakeKeyHash
+      , PaymentKeyHash_ScriptHash
+      , ScriptHash_ScriptHash
+      , PaymentKeyHash_Pointer
+      , ScriptHash_Pointer
+      , PaymentKeyHash
+      , ScriptHash
+      )
+  , addrHeaderTypeUInt
+  , addrHeaderType
+  )
+
+-- | This type class asserts that two types (native Purescript and
+-- | foreign Javascript) can be converted to each other.
+class
+  NativeForeignConvertible (native :: Type) (frgn :: Type)
+  | native -> frgn
+  , frgn -> native where
+  toNativeType :: frgn -> Maybe native
+  toForeignType :: native -> Maybe frgn
```

Comment:
I think that conversions that produce Maybes will make it tiresome to handle Nothing cases in pure code and correctly transport out the errors, but I don't have an idea on how to 'fix' it ad-hoc.
Using something like `ExceptV` or at least `type EitherV er a = Either (Variant er) a` throughout CTL apis and then providing functions for transporting errors, so  sth like:
`toNativeType :: forall r m. frgn -> ExceptV (frgnConvErr :: FrgnConvErr | r) m native`
along with handlers, "monad Identity cases" and hushed versions:
`toNativeType_ :: forall r m. frgn -> Maybe native`
looks like a solution to that in a long run.
cc @ngua

Reply:
I agree that this will be cumbersome to work with. Let's address this in #107 as well. We probably want a single way to handle such cases across the codebase